### ISSUE TITLE
[FEATURE] Update console color output to match the new color customization support in buildkit

### ISSUE THREAD

**Issue by spkane** [docker/compose#9818]
**Description**

buildkit output colors can now be configured or completely disabled. This has been merged into the currently released versions of `buildkit` and `docker buildx`, and will show up in `docker build` as soon as Docker CLI v22.06.0. ships.

See: https://github.com/moby/buildkit/pull/2954

It would be really nice to implement the same support into this plugin (and other core docker plugins that output to the console).

The current dar blue is tough to read on a dark background.

For buildkit, the env var looks something like this:

`BUILDKIT_COLORS=run=green:warning=yellow:error=red:cancel=cyan`

It probably makes sense to support a `DOCKER_COLORS` env var which is read first by a plugin and then can be overridden by a more specific one if that is set (as in the BUILDKIT case). 

It is probably worth putting all this logic into a library that all these plugins can use, but that could be done as a later re-factor as well.

**Additional information you deem important (e.g. issue happens only occasionally):**

Having constant and readable output is pretty important when pairing, mobbing, teaching etc, so making this configurable is pretty useful and avoids forcing people to change their terminal colors, just so the output can be read easily.

**Output of `docker compose version`:**

```
Docker Compose version v2.10.2-13-ge7b488bb
```

**Additional environment details:**

* I built the `compose` plugin from the `v2` to make sure there wasn't an unreleased code that addressed this.

**Comment by davebarkerxyz** [comment]
+1

Being able to disable the colours using an environment variable would be a really useful improvement. The colours are currently unreadable on my light terminal, and having to specify ‘—no-color’ on every invocation is cumbersome.

**Comment by milas** [comment]
Based on https://github.com/moby/buildkit/pull/2954#issuecomment-1190863831, this will likely need to wait until BuildKit v0.11 before support can be added to Compose.

**Comment by spkane** [comment]
@milas This change has actually already been shipped in buildkit and pulled into buildx v0.9.0 (https://github.com/docker/buildx/releases/tag/v0.9.0) and is now available in the current Docker Desktop release 4.12.0 (https://docs.docker.com/desktop/release-notes/#docker-desktop-4120).

When the Docker CLI v22.06.0 ships (there is a beta at the moment), `docker build` will also include this, due to the work in this issue: https://github.com/docker/cli/pull/3314

Adding matching functionality for the `docker compose` output should make it possible to have most of the common/important output that uses the newer style look the same.

**Comment by milas** [comment]
Yes, the versioning + dependency situation is a bit complex 😅 

Compose depends upon `moby/moby` + `docker/cli` as well as `moby/buildkit` + `docker/buildx`.

Once we've got appropriate stable versions of all those available and pulled into Compose, this should Just Work ™️ 

Sneak peek:
<img width="1274" alt="Screen Shot 2022-09-08 at 3 27 41 PM" src="https://user-images.githubusercontent.com/841263/189209217-234b866d-5f39-4774-9dde-8aafedb77257.png">

**Comment by spkane** [comment]
Makes sense, but what about the rest of the output from `compose` ?

The intention here was to suggest that all of the compose plugin output should be updated in a similar fashion so that setting something like `COMPOSE_COLORS` would do the same thing here, for things other than the build.

In the long run, I could see a higher level env var, called `DOCKER_COLORS` which is read in first by all these plugins, etc., and then could be overridden by the more specific env vars if desired, and that most of this output color logic is actually in a library somewhere that all these plugins can use.

**Comment by ndeloof** [comment]
AFAICT this is already supported with latest codebase:

https://asciinema.org/a/oJ5LmjJViKmx6COQTPEOTsTzP

**Comment by spkane** [comment]
@ndeloof This issue is about making ALL the output from the compose plugin, to support colorization. This ONLY works currently  when compose is using buildkit, it does not work for anything specific to `docker compose` like `docker compose up`, etc.

This is why the description was suggesting that it might make sense to support a more generic `DOCKER_COLORS` env var that sets the colors used for output for any code/plugin that supports it.

This is an example of the plugin using the default dark blue, which is almost unreadable with a dark background.

<img width="1033" alt="CleanShot 2023-03-04 at 09 11 58@2x" src="https://user-images.githubusercontent.com/129629/222919490-c8fcca6c-c4c0-46ed-87d4-563108773330.png">

**Comment by ndeloof** [comment]
I agree it would be nice to have `DOCKER_COLORS` to apply to all docker commands.
It would then require https://github.com/moby/buildkit/pull/2954 to be an exported codebase in docker/cli (or a dedicated library?) so all plugins can rely on it. Being non-exported codebase in buildkit and dedicated to buildx needs makes it hard to adopt without a mass code duplication and risks for future inconsistencies.  Also always has been a mystery to me buildkit includes console formatting logic and this is not buildx role to do this

**Comment by Tux** [comment]
It would be even nicer if the default color scheme would default to light mode instead of cusing light gray on white or something similar when not in dark mode so I can read neither question nor answer. Ultimately annoying.

Though dark mode may appeal to many, it also does not appeal to many. Please consider!
Better to have no colors at all than to make feedback completely useless as one cannot read it

(a light-mode user who uninstalls most utilities that only support dark mode as these are making my eyes utterly tired)

**Comment by ndeloof** [comment]
@Tux this is not really a "dark mode" choice, just legacy terminals used white (or green) on black background, so the default coloring scheme. There's unfortunately no portable way to guess terminal colors to adjust to your local settings.

**Comment by ndeloof** [comment]
not configurable yet, see https://github.com/docker/compose/issues/9818#issuecomment-1830048760

**Comment by ndeloof** [comment]
`BUILDKIT_COLORS` as name suggest is specific to BuildKit
For compose to adopt such a convention this should be driven by docker/cli and apply to all `docker xx` commands.
Closing as "not planned"

**Comment by spkane** [comment]
This request was not about using the BUILDKIT_COLORS code directly or about getting full support from the upstream Docker codebase. Although I agree this would be ideal, it would still be very nice, if there was a simple way to adjust the output colors for the plugin. Is this really so hard to implement that discussing it for 3 years and then closing the request is easier then putting in some simple stop-gap functionality until there is a broader solution for the whole ecosystem?

**Comment by ndeloof** [comment]
@spkane this is not about being hard to implement, but about where this must be discussed and implemented.
As a CLI plugin, Compose should align with other Docker CLI components. We already made some changes to offer a more homogeneous UX, relying on docker/cli config to support user-customized formats. There's still many places Compose isn't well aligned, so better not add more.
If/Once docker/cli has color settings set, I'd be happy we rely on those in Compose

**Comment by Tux** [comment]
Until then, it would be better to disable *all* colors in the CLI, because it currently just adds value to people using a dark background, which is not an option on many desktops for many reasons.
(I have even recently spotted someone using a light-red pinkish background for medical reasons with an eye problem)

**Comment by ndeloof** [comment]
@Tux my terminal let me define color profiles to map ANSI colors to actual colors I want to get displayed. Sounds to me this is a better alternative vs getting all possible CLI to allow and be configured for user-defined colors, while there's no commonly used  / POSIX approach to this issue (but  `NO_COLOR`)

**Comment by yurikhan** [comment]
> my terminal let me define color profiles to map ANSI colors to actual colors I want to get displayed.

Yes, but dark blue (ANSI color 4) has to be dark enough to be usable as a background in Midnight Commander with all other ANSI colors (with a possible exception for black) readable against that. And the system of inequalities “{ contrast_ratio(c_0, c_4) ≥ 4.5; contrast_ratio(c_4, c_i) ≥ 4.5 for i in 1..3, 5..15; colors c_0..c_15 are distinct enough from each other and from pure white }” is unsolvable.

Please just do not use dark blue as a foreground color.
